ui/src/pages/Edit/UploadStory.tsx

```diff
@@ -0,0 +1,721 @@
+        author: {
+          ID: 0,
+          CreatedAt: "",
+          UpdatedAt: "",
+          DeletedAt: {
+            Time: "",
+            Valid: true,
+          },
+          firstName: formInput.author_last_name as string,
+          currentCity: formInput.current_city as string,
+          bio: formInput.bio as string,
+        } as Author,
+        tags: [],
+        is_visible: true,
+        latitude: 0,
+        longitude: 0,
+        ID: 0,
+        CreatedAt: "",
+        UpdatedAt: "",
+        DeletedAt: {
+          Time: "",
+          Valid: true,
+        },
+      };
+      return storyFromData;
+    }
+    return undefined;
+  }, [state.isDrawerOpen, formInput]);
+
+  const hasAllRequiredFields = React.useMemo(() => {
+    return (
+      formInput.image &&
+      formInput.title &&
+      formInput.content &&
+      formInput.summary &&
+      formInput.author_first_name &&
+      formInput.author_last_name &&
+      formInput.year &&
+      formInput.current_city &&
+      formInput.author_country
+    );
+  }, [formInput]);
+
+  const setImage = (files: File[]) => {
+    setFormInput({
+      image: files[0],
+    });
+  };
+
+  const handleKeyDown = (event) => {
+    switch (event.key) {
+      case ",":
+      case " ": {
+        event.preventDefault();
+        event.stopPropagation();
+        if (event.target.value.length > 0) {
+          dispatch({
+            type: "SET_TAG_VALUES",
+            tags: [...state.tagArray, event.target.value],
+          });
+        }
+        break;
+      }
+      default:
+    }
+  };
+
+  const setErrorUploadingState = (event, reason) => {
+    if (reason === "clickaway") {
+      return;
+    }
+    dispatch({ type: "SET_ERROR_STATE", errorState: false });
+  };
+
+  const handleChange = (
+    event?: React.ChangeEvent<{ name?: string; value: unknown }>
+  ) => {
+    const name = event.target.name as keyof typeof formInput;
+    setFormInput({
+      [name]: event.target.value,
+    });
+  };
+
+  const storySubmitDialog = (result) => {
+    dispatch({ type: "SET_LOADING_STATUS", loadingStatus: false });
+    dispatch({ type: "SET_BUTTON_DISABLE", status: false });
+    const resultMessage = JSON.parse(result).message;
+    if (
+      resultMessage === "Story Added Successfully" ||
+      resultMessage === "Story Updated successfully"
+    ) {
+      //TODO: Fix redirect (currently does nothing)
+      // console.log("HERE");
+      return <Redirect to="/admin" />;
+    } else {
+      dispatch({ type: "SET_ERROR_STATE", errorState: true });
+    }
+  };
+
+  const addCountryButtonPressed = (autocompleteAuthorCountry: string) => {
+    setFormInput({
+      ["author_country"]: autocompleteAuthorCountry,
+    });
+    dispatch({ type: "ADD_COUNTRY", newCountry: autocompleteAuthorCountry });
+    dispatch({
+      type: "SET_AUTHOR_COUNTRY",
+      country: autocompleteAuthorCountry,
+    });
+    countries.push(autocompleteAuthorCountry);
```

Comment:
mutable state is icky. What we should do is have two layers.

```
// handles data fetching
function UploadStoryWrapper() {
  const { data: tagOptions, error } = useSWR<string[]>("/api/tags");
  const { data: countries, error: errorCountries } = useSWR<string[]>(
    "/api/countries"
  );

 // handle error states
 // ...

  if (tagOptions === undefined || countries === undefined) {
    return some loading indicator
  }
  
  UploadStories component guaranteed to have filled tagsOptions and countries
  return <UploadStory tagOptions={tagOptions} countries={countries};
}
```

We can then pass in `countries` into the reducer state -> `INIT_STATE(countries, currentStory)`. Instead of calling `countries.push()`, we dispatch an action to push countries:

```
{
  ...
  countries: [...state.countries, action.newCountry]
}
```

Reply:
Okay so I've come across a problem with this suggested solution: We can't return any TSX before all the hooks are declared. If we return at the `if(tagOptions === undefined ...)`, the hooks below it will be conditionally instantiated. We can't move those hooks up because parts of them rely on the state reducer (which in your suggested edit would be AFTER the if statement) --> thus we have a cycle where we are forced to conditionally create hooks.  This is also the reason the error check for the useSWR methods appears right before the final return statement. 

One solution I thought of for this: set all of our TSX to a variable and change that variable as needed, then return it at the very end of the function. Something like:
```
let returnTSX = <div/>
if(tagOptions === ....) returnTSX = <div>IF</div>;
else returnTSX=<div>ELSE</div>;
``` 
(along those lines) 
However, the idea of setting the return TSX to a variable seems very messy to me and doesn't work quite as I thought it would when I was messing around with it. 
Let me know what you think.

Reply:
No the hooks wouldn't be conditionally rendered because they would be in two different components. Essentially we have a data fetching component, and we have a view layer component.

```ts
function UploadStoryWrapper() {
  // fetch data with SWR hooks
  // if data error return error
  // if data undefined return loading
  // if data defined return <UploadStory tagOptions={tagOptions} countries={countries} />
}

...

function UploadStory({ tagOptions, countries }) {
  const [state, dispatch] = useReducer(reducer, get_init_state(countries, currentStory));
  // do whatever hook calls you want here

  // return final product to get rendered
}
```

Reply:
got it, thanks for clarifying!